Add CentOS 8 Support

jcsiadal opened the pull request:
Adds support for CentOS 8 (should also support RHEL 8, but not yet tested). There is a change in chroot creation that needs to be reviewed and approved.

List of changes and details
Update WW versions in autoconfigure to 3.9

- Update the wwinit NTP script to terminate if chrony is found. There is no "ntp" RPM in CentOS 8, so was producing an error. TODO: Add git issue to create a 51-chrony.init module.
- Include my platform build script for CentOS 8. Not intended for release, but for development. It's important to note that I needed to use more OS-provided components, as the included sources would not compile in CentOS 8. I think a good future project is to detect the target OS and do this during autoconfigure.
- Update the wwvnfs script to check for sysmacros.ph. Glibc 2.26 drops sysmacros.h, so the Perl conversion goes away in CentOS 8. I looked at different ways to handle this, since I don't know if its needed in any modern distro. My quick fix is to try and load it, and then actually load it if no error occurs. This avoids alternate scripts for different Linux versions.
- Add wwmkchroot support for CentOS 8.

This last item is a change in how the vnfs chroots are built. RHEL 8 moved to DNF, which _mostly_ supports YUM options. The current DNF really impacted operation in chroots.

The big change here was I decoupled the chroot package manager configuration from the host pkgmgr configuration. WW 3.8 copies the head node YUM config/repos to the chroot before installation, using a temporary YUM config file. I removed that. Instead, it makes the repos and configuration used during chroot installation permanent within that chroot. The head node and compute node repos remain separate. This allows me to easily use a CentOS 8 or RHEL 8 head node and build CentOS 6, 7, or 8 chroot images. I tested building all three of these.

I'll need to do some more work to make sure these cross-built chroots are maintainable on their own -- that is, updating packages in the chroot after creating it. I'd also like to confirm cross-building from SUSE or Ubuntu.

jcsiadal committed:
Merge pull request #1 from warewulf/development

Development Update

jcsiadal committed:
Update WW ver to 3.9

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal committed:
wwinit ntp: check for chrony

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal committed:
Add CentOS8 build script

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal committed:
Check for sysmacros.ph

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal committed:
First shot: CentOS8 VNFS

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal committed:
Fix missing then statements

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal committed:
Update vnfs Makefile.am

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal committed:
Add CentOS templates

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

bensallen reviewed (changes requested):
Please change the version to 3.9.0 instead of 3.9. Also the CircleCI config needs updating to get build to pass: https://github.com/warewulf/warewulf3/blob/master/.circleci/config.yml#L34. Certainly not required for this PR, but if you are so inclined it would be amazing to also have a build_centos8 job in CircleCI.

bensallen commented on cluster/libexec/wwinit/50-ntpd.init:
```diff
@@ -20,9 +20,14 @@ fi
 
 wwreqroot
 
+if  rpm -q chrony >/dev/null 2>&1 ; then
```
@jmstover Thoughts on if we should bail from 50-ntpd.init if chrony is installed? It is the default in EL7 and 8, but at least in my environment we still use ntpd everywhere. Alternatively we could reverse the order, and check for ntp.conf, if not found and chrony *is* found, exit 0 cleanly. If ntp.conf isn't found *and* chrony is not found then we exit 1.

jcsiadal replied:
Sounds fine to me. Admins shouldn't have both installed at the same time, but I just confirmed you can. Can you confirm these changes will do what you're asking? Changing the filenames so there's a clear order to wwinit.

1. 51-chrony.init: if ntp or openntpd found then return error 0.
2. 51-chrony.init: if chrony not found, return error 1.
3. 52-ntp.init: if ntp/openntpd not found, check for chrony.
4. 52-ntp.init: if chrony found, return error 0, else return error 1.

jcsiadal committed:
Add dracut workaround

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal committed:
Update version to 3.90

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal committed:
ntpd init updated

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal commented:
I updated the version # and ntpd init. In addition, I added a workaround for new Dracut to the mkchroot, if a kernel is included at build-time. Unfortunately, it doesn't help with kernel installs on a working CentOS8 chroot.

jcsiadal committed:
Merge branch 'development' into centos8

bensallen commented:
> Please change the version to 3.9.0 instead of 3.9.

@jcsiadal Looks like the version is now 3.90 instead of 3.9.0. Can you fix this to set the version to 3.9.0?

bensallen commented:
https://github.com/warewulf/warewulf3/blob/master/.circleci/config.yml#L34 also needs to be updated to install the current RPMs based on 3.9.0 to get the CircleCI builds passing.

jcsiadal committed:
Update specs and CI

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal committed:
Fix cicrlecl

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal commented:
Fixed version to 3.9.0.
Updated the rpm spec.in files as needed.
Updated the circleci config. No simple way to test it on my side, and my first attempt is broken. Since I have the error output, I'll work on getting this fixed.

jcsiadal committed:
YAML lint circleci config

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal committed:
circleci: enable PowerTools repo

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal committed:
Add DNF plugins to circleci

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal committed:
circleci: Add perl test harness

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal commented:
Leaving the PR in place right now, because I need the CI tests running. I don't know how to make that happen outside of a PR. When this is working, I'll close the PR, rebase, squash the commits, and resubmit.

jcsiadal committed:
circleci: Add perl-JSON-PP

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal committed:
Fix wwinit Makefile

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal committed:
Fix spec if tags

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal committed:
circleci: Add ipmi debug

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal committed:
circleci: Fix autogen options

Signed-off-by: jcsiadal <jeremy.c.siadal@intel.com>

jcsiadal committed:
Debug ipmitool

jcsiadal committed:
Debug ipmitool

jcsiadal committed:
circleci: Add which rpm

jcsiadal committed:
circleci: Update autogen options

jcsiadal committed:
circleci: Update autogen options

jcsiadal committed:
ipmi debug

jcsiadal committed:
more ipmi debug

jcsiadal committed:
specfile updates

jcsiadal committed:
Specfile fix

jcsiadal committed:
Fix ipmitool spec

jcsiadal committed:
Add libtirpc

jcsiadal committed:
Macro to global

jcsiadal committed:
Disable cross-compile

jcsiadal committed:
Change RPM macros

jcsiadal commented:
Replaced by #207
